Tests/Realm.Tests/Database/AsyncTests.cs

```diff
@@ -155,5 +158,405 @@ static async Task<long> MeasureTiming(Func<Task> func)
+
+                    using (var transaction = await innerRealm.BeginWriteAsync())
+                    {
+                        var person = new Person { FullName = personName };
+                        innerRealm.Add(person);
+                        Assert.That(peopleQuery.Count(), Is.EqualTo(1));
+                    }
+                }
+
+                Assert.That(_realm.All<Person>().Count(), Is.EqualTo(0));
+            });
+        }
+
+        [Test]
+        public void WriteAsync_Start_Multiple_AsyncTransaction_NoDeadlock()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                var transaction = await _realm.BeginWriteAsync();
+
+                try
+                {
+                    _realm.Add(new Person
+                    {
+                        FirstName = "Marco"
+                    });
+
+                    var writeTask = _realm.WriteAsync(() =>
+                    {
+                        _realm.Add(new Person
+                        {
+                            FirstName = "Giovanni"
+                        });
+                    });
+
+                    await transaction.CommitAsync();
+                    await writeTask;
+
+                    var peopleQuery = _realm.All<Person>();
+                    Assert.That(peopleQuery.Count, Is.EqualTo(2));
+                    Assert.That(peopleQuery.ElementAt(0).FirstName, Is.EqualTo("Marco"));
+                    Assert.That(peopleQuery.ElementAt(1).FirstName, Is.EqualTo("Giovanni"));
+                }
+                catch
+                {
+                    transaction.Dispose();
+                }
+            });
+        }
+
+        [Test]
+        public void WriteAsync_CaptureObjectsInDelegate_DoesNotThrow()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                var person = new Person
+                {
+                    FirstName = "Marco"
+                };
+
+                await _realm.WriteAsync(() =>
+                {
+                    _realm.Add(person);
+                });
+            });
+        }
+
+        [Test]
+        public void AsyncWrite_Func_RealmObj()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                var peopleQuery = _realm.All<Person>();
+                var personName = "Jonh Wickie";
+                Assert.That(peopleQuery.Count(), Is.EqualTo(0));
+
+                var person = await _realm.WriteAsync(() =>
+                {
+                    var person = new Person { FullName = personName };
+                    _realm.Add(person);
+                    return person;
+                });
+
+                Assert.That(peopleQuery.Count(), Is.EqualTo(1));
+                Assert.That(peopleQuery.First().FullName, Is.EqualTo(personName));
+                Assert.That(person, Is.Not.Null);
+                Assert.That(person.FullName, Is.EqualTo(personName));
```

Comment:
These conditions could be simplified to

```suggestion
                Assert.That(peopleQuery.Single(), Is.EqualTo(person));
```